compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIENormalizeLoopBounds.cpp

```diff
@@ -0,0 +1,191 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree-amd-aie/Transforms/Passes.h"
+#include "iree-amd-aie/Transforms/Transforms.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+
+#define DEBUG_TYPE "iree-amdaie-normalize-loop-bounds"
+
+namespace mlir::iree_compiler::AMDAIE {
+
+/// NOTE: Copy of existing utility struct in llvm-project:
+/// https://github.com/llvm/llvm-project/blob/172759492a162592da7ae9e03888661c108b1be4/mlir/lib/Dialect/SCF/Utils/Utils.cpp#L33C1-L40C15.
+/// Can be replaced with that utility struct once it is exposed.
+///
+// This structure is to pass and return sets of loop parameters without
+// confusing the order.
```

Comment:
I was about to suggest `emitNormalizedLoopBounds` from upstream but unfortunately it's not exposed. Good catch.

In fact, I strongly feel this particular pass should be upstreamed! There's no reason for it to stay as an `AMDAIE` specific pass. While the upstream merge takes time, we can definitely unblock us by getting this pass in - but we should actually consider upstreaming this.

Reply:
Yeah, I also feel like it should be upstreamed, but I wanted to get this in to unblock first and have a discussion on how to upstream in the best way.